indra/llrender/llfontfreetypesvg.cpp

```diff
@@ -0,0 +1,205 @@
+		return FT_Err_Invalid_SVG_Document;
+	}
+
+	// We don't (currently) support transformations so test for an identity rotation matrix + zero translation
+	if (document->transform.xx != 1 << 16 || document->transform.yx != 0 ||
+		document->transform.xy != 0 || document->transform.yy != 1 << 16 ||
+		document->delta.x > 0 || document->delta.y > 0)
+	{
+		datap->Error = FT_Err_Unimplemented_Feature;
+		return FT_Err_Unimplemented_Feature;
+	}
+
+	float svg_width = datap->pNSvgImage->width;
+	float svg_height = datap->pNSvgImage->height;
+	if (svg_width == 0.f || svg_height == 0.f)
+	{
+		svg_width = document->units_per_EM;
+		svg_height = document->units_per_EM;
+	}
+
+	float svg_x_scale = (float)document->metrics.x_ppem / floorf(svg_width);
+	float svg_y_scale = (float)document->metrics.y_ppem / floorf(svg_height);
+	float svg_scale = llmin(svg_x_scale, svg_y_scale);
+	datap->Scale = svg_scale;
+
+	glyph_slot->bitmap.width = floorf(svg_width) * svg_scale;
+	glyph_slot->bitmap.rows = floorf(svg_height) * svg_scale;
+	glyph_slot->bitmap_left = (document->metrics.x_ppem - glyph_slot->bitmap.width) / 2;
+	glyph_slot->bitmap_top = glyph_slot->face->size->metrics.ascender / 64.f;
+	glyph_slot->bitmap.pitch = glyph_slot->bitmap.width * 4;
+	glyph_slot->bitmap.pixel_mode = FT_PIXEL_MODE_BGRA;
+
+	/* Copied as-is from fcft (MIT license) */
+
+	// Compute all the bearings and set them correctly. The outline is scaled already, we just need to use the bounding box.
+	float horiBearingX = 0.;
+	float horiBearingY = -glyph_slot->bitmap_top;
+
+	// XXX parentheses correct?
+	float vertBearingX = glyph_slot->metrics.horiBearingX / 64.0f - glyph_slot->metrics.horiAdvance / 64.0f / 2;
+	float vertBearingY = (glyph_slot->metrics.vertAdvance / 64.0f - glyph_slot->metrics.height / 64.0f) / 2;
+
+	// Do conversion in two steps to avoid 'bad function cast' warning
+	glyph_slot->metrics.width = glyph_slot->bitmap.width * 64;
+	glyph_slot->metrics.height = glyph_slot->bitmap.rows * 64;
+	glyph_slot->metrics.horiBearingX = horiBearingX * 64;
+	glyph_slot->metrics.horiBearingY = horiBearingY * 64;
+	glyph_slot->metrics.vertBearingX = vertBearingX * 64;
+	glyph_slot->metrics.vertBearingY = vertBearingY * 64;
+	if (glyph_slot->metrics.vertAdvance == 0)
+	{
+		glyph_slot->metrics.vertAdvance = glyph_slot->bitmap.rows * 1.2f * 64;
+	}
+
+	return FT_Err_Ok;
+}
+
+// static
+FT_Error LLFontFreeTypeSvgRenderer::OnRender(FT_GlyphSlot glyph_slot, FT_Pointer*)
+{
+	LLSvgRenderData* datap = static_cast<LLSvgRenderData*>(glyph_slot->generic.data);
+	llassert(FT_Err_Ok == datap->Error);
+	if (FT_Err_Ok != datap->Error)
+	{
+		return datap->Error;
+	}
+
+	// Render to glyph bitmap
+	NSVGrasterizer* nsvgRasterizer = nsvgCreateRasterizer();
+	nsvgRasterize(nsvgRasterizer, datap->pNSvgImage, 0, 0, datap->Scale, glyph_slot->bitmap.buffer, glyph_slot->bitmap.width, glyph_slot->bitmap.rows, glyph_slot->bitmap.pitch);
+	nsvgDeleteRasterizer(nsvgRasterizer);
+	nsvgDelete(datap->pNSvgImage);
```

Comment:
I am still trying to build a mental model of how this works from the diff but is onRender(..) called many, many times and the objects are being created/destroyed accordingly here? Possible to cache them at initialization time ?

Reply:
The emoji font ends up working the same way as existing fonts: when a character is used for the first time the glyph is rendered to a font bitmap. From that point on rendering the glyph is using the font bitmap as a texture atlas.

I added a debug option to dump those bitmaps to disk (which can be helpful with invalid renders/sizing issues):
  * Develop / UI / Dump Font Textures - then look at the viewer's log folder @ %appdata%\SecondLife\logs

![Twitter Color Emoji_160_0_0_clr](https://user-images.githubusercontent.com/42094019/223279643-25d051a4-13be-4fb1-8381-fa4343377c88.png)

I thought about pre-rendering all the emojis at viewer start-up but I wasn't sure if that wasn't going to be "wasteful" on texture memory vs just rendering as they are needed.

I could add it as a debug option to experiment with?